File: 2020/day20/python/day20.py
#!/usr/bin/env python3
from os.path import dirname
from os.path import realpath
from os.path import join
import time
import pytest
import sys
import re

sys.path.append(join(dirname(realpath(__file__)), *["..", ".."]))

from jiggsaw import Jiggsaw


def transform_input(input_):
    # custom transform for the day
    image = dict()

    # Go through each line
    for inp in input_.splitlines():
        # Skip empty line
        if not inp:
            continue
        # Check for tile number
        tile = re.search("(\\d+)", inp)
        if tile:
            tile_nr = tile[0]
            image[tile_nr] = []
        # If no tile number, append the row to the current tile
        else:
            image[tile_nr].append(inp)

    return image


def read_input(file_name="../data/input.txt"):
    dir_path = dirname(realpath(__file__))
    with open(join(dir_path, file_name), "r") as f:
        input_ = transform_input(f.read())

    return input_


def solve_part1(input_):
    # Feed the Jiggsaw class all image pieces
    jigsaw = Jiggsaw(input_)
    jigsaw.solve()

    prod = 1
    for pos in jigsaw.get_corners():
        prod *= int(pos)
    return prod


def solve_part2(data):

    jiggsaw = Jiggsaw(data)
    jiggsaw.solve()

    monster = [
        "                  # ",
        "#    ##    ##    ###",
        " #  #  #  #  #  #   ",
    ]

    return jiggsaw.get_rough_waters(monster)


def main():
    input_ = read_input()

    t0 = time.time()
    part1 = solve_part1(input_)
    time_part1 = round((time.time() - t0) * 1e3)
    print(f"Solution to part one: {part1} (time taken {time_part1}[ms])")

    t0 = time.time()
    part2 = solve_part2(input_)
    time_part2 = round((time.time() - t0) * 1e3)
    print(f"Solution to part two: {part2} (time taken {time_part2}[ms])")


if __name__ == "__main__":
    main()


@pytest.mark.parametrize(
    "input1, output1", [("../data/test_input0.txt", 20899048083289)]
)
def test1(input1, output1):
    assert solve_part1(read_input(input1)) == output1


@pytest.mark.parametrize("input2, output2", [("../data/test_input0.txt", 273)])
def test2(input2, output2):
    assert solve_part2(read_input(input2)) == output2

File: 2020/day20/cpp/src/day20.cpp
#include "2020/day20/cpp/include/day20.hpp"

#include <algorithm>
#include <map>
#include <queue>
#include <unordered_map>
#include <unordered_set>

#include "2020/day20/cpp/include/jiggsaw.hpp"
#include "utils/cpp/include/utils.hpp"
using day20::input_type;
using day20::output_type;

input_type day20::transform_input(std::vector<std::string> &input) {
  input_type image;
  std::string tile_nr;

  // Go through each line
  for (auto &inp : input) {
    // Skip empty line
    if (inp.length() == 0) {
      continue;
    }
    // Check for tile number
    if (inp[0] == 'T') {
      tile_nr = inp.substr(5, inp.length() - 6);
      // If no tile number, append the row to the current tile
    } else {
      image[tile_nr].push_back(inp);
    }
  }

  return image;
}

input_type day20::read_input(std::string file_name) {
  auto input = utils::read_file(file_name);
  return transform_input(input);
}

// Implement part one solution
output_type day20::solve_part1(const input_type &input_) {
  // Feed the Jiggsaw class all image pieces
  Jiggsaw j(input_);
  j.solve();

  uint64_t prod(1);
  for (auto &key : j.get_corners()) {
    prod *= std::stoll(j.m_complete_images[key].first);
  }
  return prod;
}

// Implement part two solution
output_type day20::solve_part2(const input_type &input_) {
  // Feed the Jiggsaw class all image pieces
  Jiggsaw j(input_);
  j.solve();

  std::vector<std::string> monster(
      {"                  # ", "#    ##    ##    ###", " #  #  #  #  #  #   "});
  return j.get_rough_waters(monster);
}
